src/Bicep.Core/Tracing/AzureEventSourceListenerFactory.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Azure.Core.Diagnostics;
+using Bicep.Core.Features;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+
+namespace Bicep.Core.Tracing
+{
+    public static class AzureEventSourceListenerFactory
+    {
+        public static AzureEventSourceListener Create(TraceVerbosity verbosity) => new((eventArgs, formattedMessage) => WriteTraceLine(eventArgs, formattedMessage, verbosity), EventLevel.LogAlways);
+
+        private static void WriteTraceLine(EventWrittenEventArgs eventArgs, string formattedMessage, TraceVerbosity verbosity)
+        {
+            var updatedMessage = GetFormattedMessageWithoutHeaders(
+                verbosity,
+                eventArgs.EventSource.Name,
+                eventArgs.Message,
+                eventArgs.PayloadNames,
+                eventArgs.Payload) ?? formattedMessage;
+
+            Trace.WriteLine(string.Format(CultureInfo.InvariantCulture, "[{0}] {1}", eventArgs.Level, updatedMessage), eventArgs.EventSource.Name);
+        }
+
+        // EventWrittenEventArgs does not appear mockable, so we will expose this method with deconstructed parameters
+        public static string? GetFormattedMessageWithoutHeaders(TraceVerbosity verbosity, string eventSourceName, string formatString, ReadOnlyCollection<string> parameterNames, ReadOnlyCollection<object> parameterValues)
```

Comment:
>GetFormattedMessageWithoutHeaders [](http://example.com/codeflow?start=30&length=33)

Other options considered:
1. Logging all the headers all the time - overwhelming amount of mostly useless data in the trace spew.
2. Logging first line only - does not cover most cases. Some events have multi line strings without logging headers.
3. Tight coupling with [https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs) - would require continuous updates to match changes there.

We didn't want to do option 1 for default tracing (which require opt-in, btw), so the current implementation seems like the most maintainable option with fewest issues. Definitely open to better ideas, though.

Reply:
Testing this is also tricky because the `EventWrittenEventArgs` doesn't appear to be mockable as far as I can tell.